crates/ra_hir/src/expr.rs

```diff
@@ -792,40 +795,79 @@ impl ExprCollector {
             ast::ExprKind::Label(_e) => self.alloc_expr(Expr::Missing, syntax_ptr),
             ast::ExprKind::IndexExpr(_e) => self.alloc_expr(Expr::Missing, syntax_ptr),
             ast::ExprKind::RangeExpr(_e) => self.alloc_expr(Expr::Missing, syntax_ptr),
-            ast::ExprKind::MacroCall(_e) => self.alloc_expr(Expr::Missing, syntax_ptr),
+            ast::ExprKind::MacroCall(e) => {
+                let name = e.name().map(|nr| nr.as_name()).unwrap_or_else(Name::missing);
+
+                if let Some((module, _)) = self.resolver.module() {
+                    let def = module.find_macro(&name);
+
+                    if let Some(def_id) = def {
+                        let rules = db.macro_def(*def_id);
+
+                        if let Some(rules) = rules {
+                            let tt = e.token_tree().map(mbe::ast_to_token_tree);
+
+                            if let Some(Some((tt, _))) = tt {
+                                let result = rules.expand(&tt);
+
+                                if let Ok(tt) = result {
+                                    // let items = ;
+
+                                    mbe::token_tree_to_ast_item_list(&tt)
+                                        .syntax()
+                                        .descendants()
+                                        .find_map(ast::Block::cast)
```

Comment:
Yeah, the problem here is that we only expand macros to list of items, we don't support expanding macro to an expression yet... This should be fixable though!

Reply:
Do you think a `token_tree_to_syntax_node` is enough in here? Or a `token_tree_to_syntax_block` ?

Reply:
I think some extension of interface is required. 

IIRC, in rustc there's a bunch of functions like

* expand as items
* expand as expressions
* expand as statements

etc.

That is, the caller decides what macro expands to. I *think* this should be encoded as a part of `MacroCallId`. That is, `MacroLoc` should include an enum which says if this is an item-position macro, an expr position macro, etc

Reply:
I will propose we introduce a new enum `MarcoKind` in mbe, which represents the type of macro want to expand to. And a `token_tree_to_macro_kind` would convert a subtree to `MacroKind`. 

On the other hand, what happen if a subtree could not match the provided `MacroKind` ? For example, a `MacroKind::Block` was given and the subtree itself are `statements` ? Could `token_tree_to_macro_kind` should convert it to (i.e, add a Delimiter::brace in the root node of subtree) ? Or it should just return a `Result` ?

Reply:
Yes, `MacroKind` seems like a good API!

If the expansion can not be parsed as a specified kind, we should return an `Err`.

Reply:
Created a PR for that, after it is merged, @Lapz could use that instead.